### PR TITLE
Use refreshable for runtime config

### PR DESCRIPTION
**Pull request by jkozlowski**
**Goals (and why)**:

Should allow us to remove CachedComposingSupplier class (in separate PR) and simplify diff in https://github.com/palantir/atlasdb/pull/4706.

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    static ConfigRefreshable createPolling(Supplier<Optional<AtlasDbRuntimeConfig>> config) {
+        DefaultRefreshable<Optional<AtlasDbRuntimeConfig>> refreshable = new DefaultRefreshable<>(call(config));
+
+        ScheduledExecutorService executor = PTExecutors.newSingleThreadScheduledExecutor();
+        executor.scheduleWithFixedDelay(
+                () -> {
+                    try {
+                        refreshable.update(config.get());
+                    } catch (Throwable e) {
+                        // This should not occur in practice
```
Comment (jkozlowski) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java:73]: This isn't ideal, but if this occurs the application is doing something really rather bad.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 2]
```diff
@@ -202,10 +202,13 @@
...
-    Supplier<Optional<AtlasDbRuntimeConfig>> runtimeConfigSupplier() {
-        return Optional::empty;
-    }
+    abstract Optional<Refreshable<Optional<AtlasDbRuntimeConfig>>> runtimeConfig();
+
+    /**
+     * @deprecated use {@link #runtimeConfig} instead.
+     */
+    @Deprecated
+    abstract Optional<Supplier<Optional<AtlasDbRuntimeConfig>>> runtimeConfigSupplier();
```
Comment (jkozlowski) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:211]: This change should be fine, the autogenerated immutables class should still have the same #runtimeConfigSupplier(Supplier) method, so should not be a break for users.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 3]
```diff
@@ -279,6 +282,12 @@ LockWatchingCache lockWatchingCache() {
         thread.start();
     };
 
+    @Value.Check
+    protected void check() {
+        Preconditions.checkState(!(runtimeConfigSupplier().isPresent() && runtimeConfig().isPresent()),
+                "Either Refreshable or Supplier of runtime config must be provided or none, but not both");
```
Comment (jkozlowski) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:288]: This maintains the current setup, where it's possible to not provide the runtime config (used for InMemory configuration).

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 4]
```diff
@@ -375,11 +388,11 @@ private TransactionManager serializableInternal(@Output List<AutoCloseable> clos
 
             // Even if sweep queue writes are enabled, unless targeted sweep is enabled we generally still want to
             // at least retain the option to perform background sweep, which requires updating the priority table.
-            if (!targetedSweepIsFullyEnabled()) {
+            if (!targetedSweepIsFullyEnabled(config(), runtime)) {
                 kvs = SweepStatsKeyValueService.create(kvs,
                         new TimelockTimestampServiceAdapter(lockAndTimestampServices.timelock()),
-                        Suppliers.compose(SweepConfig::writeThreshold, sweepConfig::get),
-                        Suppliers.compose(SweepConfig::writeSizeThreshold, sweepConfig::get)
+                        sweepConfig.map(SweepConfig::writeThreshold),
```
Comment (jkozlowski) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:394]: I removed all uses of Suppliers.compose in this class.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 2 above)
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:211]: See test in TransactionManagers I added, it's using this method.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    public Refreshable<Optional<AtlasDbRuntimeConfig>> refreshable() {
+        return delegate;
+    }
+
+    @Override
+    public void close() {
+        closer.run();
+    }
+
+    static ConfigRefreshable wrap(Refreshable<Optional<AtlasDbRuntimeConfig>> delegate) {
```
Comment (ferozco) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java:59]: Seems kinda odd to wrap the refreshable in this way. I think it was expose a way to convert a supplier to an optional using some executor, and handle the lifecycle of the executor seperately

Comment (ferozco) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java:59]: The suggestion is to just get rid of the extra level of abstraction and manage the lifecycle of the executor used to update the refreshables separately. This is especially beneficial since it allows you to reuse the executor across refreshables and allows you to avoid unnecessarily wrapping existing refreshables.

Hence:
```
ConfigRefreshable<T> config1 = ConfigRefreshable.createPolling(mySupplier1);
ConfigRefreshable<T> config2 = ConfigRefreshable.createPolling(mySupplier2);
...

config1.close();
config2.close();
```

becomes:
```
ScheduledExecutorService refreshableExecutor = PTExecutors.newSingleThreadScheduledExecutor();

Refreshable<T> config1 = Refreshables.create(mySupplier1, refreshableExecutor);
Refreshable<T> config2 = Refreshables.create(mySupplier2, refreshableExecutor);
...

refreshableExecutor.shutdown()
```

Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java:59]: I don’t expect to be using refreshables anywhere else. Handling this separately in TransactionManagers will be messy, because you don’t create the executor if someone provides the Refreshable. I can see what it would look like though

Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java]: So I tried to come up with something that's not messy and I can't: the fact that you only sometimes create the executor, and then you need to add it to the closeables list AND THEN, but only after TransactionManager is created, call #shutdownCallbacks with it makes me not want to do it.

This is actually a nice small abstraction of a closeable refreshable, and everything falls through nicely.

Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java]: Instead, I don't expose any special wrap methods, but instead the only way to create this special refreshable is from the TransactionManagers builder.

Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java]: And if we excavate this so we can remove the deprecated method, we'll just remove all this code.

Comment (ferozco) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java]: sure, whatever you're happy with

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    private static AtlasDbRuntimeConfigRefreshable createPolling(Supplier<Optional<AtlasDbRuntimeConfig>> config) {
+        DefaultRefreshable<Optional<AtlasDbRuntimeConfig>> refreshable = new DefaultRefreshable<>(call(config));
+
+        ScheduledExecutorService executor = PTExecutors.newSingleThreadScheduledExecutor();
+        executor.scheduleWithFixedDelay(
+                () -> {
+                    try {
+                        refreshable.update(config.get());
+                    } catch (Throwable e) {
+                        // This should not occur in practice
```
Comment (jkozlowski) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java:87]: Famous last words: the danger here is that we write some sort of weird update code (since DefaultRefreshable.update propagates to all the children) and this throws, and we won't find out about it. This is a bit of a change in behavior, but I don't feel strongly that this is such high risk.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:211]: I guess my bigger concern is if someone calls `txMgr.runtimeConfigSupplier()` though a code search suggests that no one does so. 

If someone did (especially if that someone is one of our friendly supporting libs) that would be a spicy break 🔥but thankfully looks like that's not the case.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:288]: nit: probably enough to just say `Cannot provide both Refreshable and Supplier of runtime config`?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 7]
```diff
@@ -538,9 +550,11 @@ TransactionConfig withConsolidatedGrabImmutableTsLockFlag(TransactionConfig tran
...
     }
 
-    private boolean targetedSweepIsFullyEnabled() {
-        return config().targetedSweep().enableSweepQueueWrites()
-                && runtimeConfigSupplier().get().map(config -> config.targetedSweep().enabled()).orElse(false);
+    private static boolean targetedSweepIsFullyEnabled(
+            AtlasDbConfig installConfig,
+            Supplier<AtlasDbRuntimeConfig> runtime) {
+        return installConfig.targetedSweep().enableSweepQueueWrites()
+                && runtime.get().targetedSweep().enabled();
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:557]: I think you drive-by fixed a bug here! Previously if you don't specify a runtime config, targeted sweep is enabled _in practice_, but the check here thinks it is disabled. Now it correctly recognises that targeted sweep is enabled by default.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                    }
+                },
+                REFRESH_INTERVAL.toNanos(),
+                REFRESH_INTERVAL.toNanos(),
+                TimeUnit.NANOSECONDS);
+
+        return new AtlasDbRuntimeConfigRefreshable(refreshable, () -> {
+            if (!MoreExecutors.shutdownAndAwaitTermination(
+                    executor, GRACEFUL_SHUTDOWN.toMillis(), TimeUnit.MILLISECONDS)) {
+                log.error("Executor did not terminate within graceful shutdown duration");
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java:98]: nit: I'd prefer `warn`, this is probably not super actionable for us?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java (hunk 6 above)
Comment (jeremyk-91) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java:87]: Not sure I follow: do we mean a subscriber? If a subscriber throws it does look like the world could be in an inconsistent state, but right now I think we're fine

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 2 above)
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:211]: Hmmm, I don't think runtimeConfigSupplier() would've compiled before? immutables shouldn't generate it, it only generates them when the type is Optional of something.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java (hunk 6 above)
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbRuntimeConfigRefreshable.java:87]: Yes, I mean that Refreshables remember everyone who called #map on them and propagate the updates. Yes you could be in an inconsistent state, but I think this is by design of refreshables.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 7 above)
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:557]: Yep

File: atlasdb-config/src/main/java/com/palantir/atlasdb/debug/TransactionPostMortemRunner.java [hunk 9]
```diff
@@ -69,7 +70,7 @@ public TransactionPostMortemRunner(
             TransactionManager transactionManager,
             TableReference tableReference,
             AtlasDbConfig install,
-            Supplier<AtlasDbRuntimeConfig> runtime,
```
[the remark this answers is not in the document]
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/debug/TransactionPostMortemRunner.java:72]: This is only instantiated in test code from what I can see? So it didn't make sense to keep it this way, and it was calling ServerListConfigs.parseInstallAndRuntimeConfigs, which now takes a Refreshable.

[the remark this answers is not in the document]
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/debug/TransactionPostMortemRunner.java:72]: Ah, nevermind, this is called in an internal product at runtime. Ok, lemme rejig this

[the remark this answers is not in the document]
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/debug/TransactionPostMortemRunner.java:72]: I'll change into Refreshable and we'll need to fixup internal product to do this properly.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 3 above)
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:288]: Done

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java (hunk 1 above)
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/ConfigRefreshable.java]: Or we are really.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/debug/TransactionPostMortemRunner.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jkozlowski) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/debug/TransactionPostMortemRunner.java:72]: Donezo
